### ISSUE TITLE
Proposal: Add functions to extract the type/permissions from an existing `FileInfo`

### ISSUE THREAD

**Issue by joshi-monster** [bcpeinhardt/simplifile#35]
Hi again :3

This is the other thing I frequently do, and that I think is generally a better pattern. simplifile currently exposes `file_info` to get the stat object, as well as a bunch of other helpers like `is_file`, etc. which all effectively do a stat syscall under the hood again, just to extract that one piece of information.

Syscalls can be quite expensive, especially if you are recursively walking directories, so as an alternative I have found myself to instead use the `mode` field and bitwise operations to extract the information myself. This means that instead of having do to 3+ stat calls (`file_info`, `is_directory`, `is_file`, ...), I only need to do one, but it can be little bit finicky to get right. l did this twice now, in both my file watcher library and also in the embeds package.

I propose adding those helper functions (`file_info_type`, `file_info_permissions`, `file_info_permissions_octal`), extracting the file type and permissions from an existing `FileInfo` object.

I've also noticed that there doesn't currently seem to be a way to get the permissions of a file? And for symmetry reasons, it might also be nice to have a `file_type` function that works like `file_info_type` but takes a path instead?

I don't have timings right now, but I can try and measure how big of a difference this would actually make in the file watcher.

I'd be happy to make a PR with those functions if you think they'd be a good addition :)
~yoshi

**Comment by bcpeinhardt** [comment]
I agree with all of this. This library started with four functions (read, write, append, delete) and grew as people asked for things, if somethings missing it means nobodies asked for it not that it shouldn't be there :) 
I wonder if rather than helper functions these should just be fields on the FileInfo object for convenience.

**Comment by joshi-monster** [comment]
The extra fields is how Erlang already does it (with `access` and `type`), Node exposes utility functions on the object instead, but it works pretty similar overall (`isDirectory` et al). I thought extra functions might be easier to add than fields, since adding a field would be a breaking change.

I do agree that in the future, doing things differently might make more sense.

I'll be happy to prepare something in the next few days, thank you! :)

**Comment by joshi-monster** [comment]
I made a simple benchmark by watching a medium-sized directory (1137 files, 260 directories) using my library; I used that one because I already had the setup and ffi code there :slightly_smiling_face: 

I replaced the `lstat` implementation with one that calls `lstat` twice for files and symlinks (once for `is_file` and once for `file_info` afterwards), and 3 times for directories (to simulate calling `is_directory` in the else branch).

Using a single lstat, it takes 70ms per tick
The patched version with 2-3 lstat calls takes 100ms instead.

So this would be ~30% improvement for recursively walking directories shaped like Gleam projects; keep in mind that these timings also include all the code necessary to compare the old and new state of the trees, but I expect it to be similar or slightly larger for `get_files`.
